proposals/0127-get-sysvar-syscall.md

```diff
@@ -0,0 +1,236 @@
+could then be made available to BPF programs through an interface around raw
+pointers, which loads the data similarly to the interface proposed herein.
+
+This approach would depend on direct mapping of host memory to VM memory,
+which has been in development for some time and requires further research
+and development to introduce safely to the VM ABI. This extensive R&D period
+would mean Core BPF program initiatives would remain blocked.
+
+Without direct mapping, sysvar data would be fully copied to this memory
+region each time a VM is created (currently one per transaction processed),
+causing a degradation in runtime performance and likely reducing compute
+by far less compared to the compute savings posed by this proposal.
+
+Even with direct mapping, this change makes address translation more complex
+and therefore slows down _all_ memory accesses.
+
+## New Terminology
+
+N/A.
+
+## Detailed Design
+
+A single `sol_get_sysvar` syscall can be defined, which can be used by all
+forward-facing sysvar interfaces to retrieve data from any sysvar.
+
+```c
+/**
+ * Retrieves a slice of data from a sysvar and copies it to VM memory.
+ *
+ * @param sysvar_id The identifier of the sysvar to retrieve data from.
+ * @param var_addr  VM memory address to copy the retrieved data to.
+ * @param offset    The offset to start copying data from.
+ * @param length    The length of data to copy, in bytes.
+ * @return          A 64-bit unsigned integer error code:
+ *                    - 0 if the operation is successful.
+ *                    - Non-zero error code.
+ *
+ * If the operation is not successful, data will not be written to the
+ * provided VM memory address.
+ */
+uint64_t sol_get_sysvar(
+  /* r1 */ void const * sysvar_id,
+  /* r2 */ uint8_t *    var_addr,
+  /* r3 */ uint64_t     offset,
+  /* r4 */ uint64_t     length,
+);
+
+```
+
+### Control Flow
+
+The syscall aborts the virtual machine if any of these conditions are true:
+
+- Not all bytes in VM memory range `[sysvar_id, sysvar_id + 32)` are readable.
+- Not all bytes in VM memory range `[var_addr, var_addr + length)` are writable.
+- `offset + length` is not in `[0, 2^64)`.
+- `var_addr + length` is not in `[0, 2^64)`.
+- Compute budget is exceeded.
+
+All VM violations (above) are checked first, before to returning one of the
+following error codes. The following checks are completed in the order they
+appear below.
+
+The syscall returns the following graceful codes:
+
+- `2` if the sysvar data is not present in the Sysvar Cache.
+- `1` if `offset + length` is greater than the length of the sysvar data.
+- `0` if the process completed successfully and the requested sysvar data was
+  written to the virtual memory at `var_addr`.
+
+### Compute Unit Usage
+
+The syscall will always consume the same amount of CUs regardless of
+control flow.
+
+CU usage is proportional to the `length` parameter.
+
+```
+sysvar_base + (32 / cpi_per_u) + max(mem_op_base, (length / cpi_per_u))
```

Comment:
Sorry for the flyby here, but I was looking at this implementation in Agave and was curious about the `(32 / cpi_bytes_per_unit)` term here. `cpi_bytes_per_unit` is currently set to `250` on mainnet, so that factor will always be 0.

It's not a big deal, and thankfully it's explicitly defined, but was there a reason to keep it separate rather than combining it with `length`? Ie instead of doing:

```
sysvar_base + max(mem_op_base, ((length + 32) / cpi_per_u))
```

Reply:
The intention was to apply `cpi_per_u` to the 32-byte sysvar ID, then to the length (or mem_op_base, whichever is max). So with that being said, what you've suggested would probably make more sense. Should be adjust this?

Reply:
Since this syscall is part of 2.0, which is really close to mainnet, it might be a bit late to do a quick change. That would probably require an amendment SIMD + new feature gate, which sounds like a lot of work for what's essentially a rounding difference of 1 CU on a syscall :sweat_smile: I think we can leave as is for now. Thanks for the response!